### DISCUSSION TITLE
How to checkout recursively?

### DISCUSSION THREAD

**Discussion by sarisia** [community/community#25429] [Codespaces]
<p>Creating Codespaces automatically checkouts repository, but not for<br>
submodules.<br>
Are there any options to do this?</p>

**Comment by acangialosi** [comment]
<p>There is an extension setting in VS Code that will clone submodules.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/3/b/3b2d4cb4ffdda957bc77e6cce87ff75f5eb77bdb.png" data-download-href="/uploads/short-url/8rvdW3yCXTqS1wxiRQZMaHaPqdB.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/5457010/181091076-bfe1dfe3-967a-4fea-a5db-42460e4c516d.png" alt="image" data-base62-sha1="8rvdW3yCXTqS1wxiRQZMaHaPqdB" width="690" height="305" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/3/b/3b2d4cb4ffdda957bc77e6cce87ff75f5eb77bdb_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">931×412 24.6 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by sarisia** [comment]
<p>Thank you for answering!<br>
I didn’t even noticed there’s a Codespace configuration in VSCode Settings.</p>
<p>By the way, I turned on <code>Codespaces: Enable Submodule Cloning</code> option and tried<br>
creating new Codespace but while the option was synced to the new Codespace properly<br>
, submodules weren’t cloned as well.</p>
<p>I have a problem with this repository: <a href="https://github.com/note-sarisia-cc/source" rel="noopener nofollow ugc">note-sarisia-cc/source</a><br>
<code>.gitmodules</code>:</p>
<pre><code class="lang-auto">[submodule "themes/zozo"]
	path = themes/zozo
	url = https://github.com/note-sarisia-cc/hugo-theme-zozo.git
</code></pre>

**Comment by jkeech** [comment, marked as answer]
<p>Hi <a class="mention" href="/sarisia">@sarisia</a>, <a class="mention" href="/acangialosi">@acangialosi</a>,</p>
<p>That setting in VS Code is only supported in Visual Studio Codespaces and is not currently implemented for GitHub Codespaces.</p>
<p>There is not currently a way to request submodule checkout in GitHub Codespaces, although it’s on our backlog! For now, you could try working around it by specifying a <code>postCreateCommand</code> in devcontainer.json to clone the submodules. Note that this will also only work today if those submodule repos are public, as the GitHub token that the Codespace gets is restricted just to the main Codespace repo.</p>

**Reply by josecelano** [reply]
I'm trying to use this solution with a public repo but I get a permission error:

```text
Please make sure you have the correct access rights
and the repository exists.
fatal: clone of 'git@github.com:torrust/torrust-index-frontend.git' into submodule path '/workspaces/torrust-codespace/index-frontend' failed
Failed to clone 'index-frontend'. Retry scheduled
Cloning into '/workspaces/torrust-codespace/tracker'...
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.
```

**Reply by oysteijo** [reply]
@josecelano : I have the same problem. Did you find a solution?

**Reply by josecelano** [reply]
> @josecelano : I have the same problem. Did you find a solution?

No, I didn't. It was only a test and I didn't continue working on that.

**Comment by sarisia** [comment]
<p>Thank you for providing details!</p>
<p>As a workaround, I added:</p>
<pre><code class="lang-json">"postCreateCommand": "git submodule update --init",
</code></pre>
<p>to <code>devcontainer.json</code> and it worked well.</p>

**Comment by jessfraz** [comment]
For private repos you can do the following:

If you follow the instructions here for adding permissions for repos:  https://docs.github.com/en/codespaces/managing-your-codespaces/managing-repository-access-for-your-codespaces be sure to add `content: read` for the specific submodule you repos

Then you can add the following to your devcontainer.json to checkout submodules and rewrite git paths:
```
"updateContentCommand": "git config --global url.https://x-access-token:${GITHUB_TOKEN}@github.com/.insteadOf ssh://git@github.com/ && git config --global url.https://x-access-token:${GITHUB_TOKEN}@github.com/.insteadOf git@github.com/ && git config --global url.https://x-access-token:${GITHUB_TOKEN}@github.com/.insteadOf git@github.com: && git submodule update --init --recursive",
```
